File: Week_03/G20190343020055/powx-n.go
package week03

func myPow(x float64, n int) float64 {

	if n < 0 {
		n = -n
		x = 1 / x
	}

	return fastPow(x, n)
}

func fastPow(x float64, n int) float64 {

	if n == 0 {
		return 1.0
	}

	half := fastPow(x, n/2)

	if n%2 == 0 {
		return half * half
	}

	return half * half * x
}

File: Week_03/G20190343020121/powx-n.java
class Solution {
    public double myPow(double x, int n) {
        long N = n;
        if (N < 0) {
            x = 1 / x;
            N  = -N;
        }
        return fastPow(x, N);
        
    }
    public double fastPow(double x, long n) {
        if (n == 0) {
            return 1;
        }
        double half = fastPow(x,n / 2);
        if (n % 2 == 0) {
            return half * half;
        } else  {
            return half * half * x;
        }
    }
}